src/build.rs

```diff
@@ -263,3 +265,323 @@ fn build_nasm(root: &Path, vendor_dir: &Path, out_dir: &Path, target_arch: &str,
+            .filter(|s| !s.trim().is_empty())
+            .map(|s|{
+                let mut s = s.split(':');
+                let arg_name = s.next().unwrap();
+                let ty_name = s.next().unwrap();
+                (arg_name.trim().to_string(), ty_name.trim().to_string())
+            }).collect();
+        let ret_ty = decl.rfind("->").map(|s| decl[s+2..].trim().to_string());
+        decls.push( Decl { name, is_pub, doc, attrs, args, ret_ty, lifetime });
+    }
+
+    decls
+}
+
+fn generate_cxx_wrappers(decls: &Vec<Decl>) -> Vec<String> {
+    decls.iter().map(|d| generate_cxx_wrapper(d)).collect()
+}
+
+fn generate_rust_wrappers(decls: &Vec<Decl>) -> Vec<String> {
+    decls.iter().map(|d| generate_rust_wrapper(d)).collect()
+}
+
+fn generate_rust_wrapper(decl: &Decl) -> String {
+    format!(r#"
+    {doc}
+    {attrs}
+    {pub_vis} unsafe fn {wrapper_name}{lifetime}({params}) -> {ret_ty} {{
+        extern "C" {{
+            fn {c_wrapper_name}{lifetime}({params}) -> WrapperResult<{ret_ty}>;
+        }}
+        let result = {c_wrapper_name}({arg_ids});
+        if result.discriminant == 0 {{
+           result.value.value
+        }} else {{
+           let WrapperError{{ buf, len, cap }}: WrapperError = result.value.error;
+           panic!(String::from_raw_parts(buf, len, cap))
+        }}
+    }}
+    "#,
+            doc = decl.doc.join("\n"),
+            attrs = decl.attrs.join("\n"),
+            pub_vis = if decl.is_pub { "pub" } else { "" },
+            wrapper_name = decl.name,
+            lifetime = decl.lifetime.clone().map(|l| format!("<{}>", l)).unwrap_or_default(),
+            params = decl.args.iter().map(|(id, ty)| format!("{}: {}", id, ty)).collect::<Vec<_>>().join(","),
+            c_wrapper_name = format!("{}_cxx", decl.name),
+            arg_ids = decl.args.iter().map(|(id, _)| id.clone()).collect::<Vec<_>>().join(","),
+            ret_ty = decl.ret_ty.clone().unwrap_or("()".to_string()),
+    )
+}
+
+fn generate_cxx_wrapper(decl: &Decl) -> String {
+    let ret_ty = decl.ret_ty.clone().map(|t| c_type(&t)).unwrap_or("void".to_string());
+    let arg_ids = decl.args.iter().map(|(id, _)| id.clone()).collect::<Vec<_>>().join(",");
+    let value_return = if decl.ret_ty.is_some() {
+        format!(
+            "return WrapperResult<{ret_ty}>::value({wrapper_name}({arg_ids}));",
+            ret_ty = ret_ty,
+            wrapper_name = decl.name,
+            arg_ids = arg_ids
+        )
+    } else {
+        format!(
+            "{wrapper_name}({arg_ids}); return WrapperResult<{ret_ty}>::value();",
+            ret_ty = ret_ty,
+            wrapper_name = decl.name,
+            arg_ids = arg_ids
+        )
+    };
+    format!(r#"
+    extern "C" WrapperResult<{ret_ty}> {c_wrapper_name}({params}) {{
+        try {{
+             {value_return}
+        }} catch(mozjpeg_rust_wrapper_exception const& e) {{
+             return WrapperResult<{ret_ty}>::error(e.error);
+        }}
```

Comment:
This will only catch and properly propagate the Rust exceptions. 

You will probably want to add a `catch(...) { std::terminate() }` that aborts the process if `mozjpeg` tries to throw anything else, because that would unwind into Rust with completely unexpected consequences (Rust will think its a Rust panic, but it isn't).